### PR TITLE
REST client for creating batch with uploading resource file

### PR DESCRIPTION
**Pull request by bowenliang123**
### _Why are the changes needed?_

- REST client support for creating batch with uploading resource file (https://kyuubi.readthedocs.io/en/master/client/rest/rest_api.html#post-batches-with-uploading-resource) introduced in #4144 

### _How was this patch tested?_
- [ ] Add some test cases that check the changes thoroughly including negative and positive cases if possible

- [ ] Add screenshots for manual tests if appropriate

- [x] [Run test](https://kyuubi.readthedocs.io/en/master/develop_tools/testing.html#running-tests) locally before make a pull request

### CODE REVIEW

File: kyuubi-rest-client/src/main/java/org/apache/kyuubi/client/IRestClient.java [hunk 1]
```diff
@@ -27,6 +28,8 @@ public interface IRestClient extends AutoCloseable {
 
   <T> T post(String path, String body, Class<T> type, String authHeader);
 
+  <T> T post(String path, HttpEntity requestEntity, Class<T> type, String authHeader);
```
[the remark this answers is not in the document]
Comment (pan3793) [reply kyuubi-rest-client/src/main/java/org/apache/kyuubi/client/IRestClient.java:31]: And it's OK if this is not intended to be used by the user directly. cc @turboFei @ulysses-you

[the remark this answers is not in the document]
Comment (ulysses-you) [reply kyuubi-rest-client/src/main/java/org/apache/kyuubi/client/IRestClient.java:31]: how about `Map<String, File>resourceFiles` ?

[the remark this answers is not in the document]
Comment (bowenliang123) [reply kyuubi-rest-client/src/main/java/org/apache/kyuubi/client/IRestClient.java:31]: But a map of key-to-file is not enough to represent the whole request entity.
In this client request,
- entity body (in media type of `multipart-formdata`)
	- subpart1 (key, string value, media type of `application-json`)
	- subpart2 (key, File, media type of file itself)
	- subpart3 (key, value, other media type) and etc.

[the remark this answers is not in the document]
Comment (ulysses-you) [reply kyuubi-rest-client/src/main/java/org/apache/kyuubi/client/IRestClient.java:31]: is these required ? I can not find them in the user-faced api `BatchRestApi`.

[the remark this answers is not in the document]
Comment (bowenliang123) [reply kyuubi-rest-client/src/main/java/org/apache/kyuubi/client/IRestClient.java:31]: End-user is using `BatchRestApi` to use this feature and they won't be concerned about how to assemble the `HttpEntity` which is done at lower level in `IRestClient`.

[the remark this answers is not in the document]
Comment (ulysses-you) [reply kyuubi-rest-client/src/main/java/org/apache/kyuubi/client/IRestClient.java:31]: I mean I do not see any code in this pr about your requirements list.. I just see the name and file. So I think name and file is enough.

[the remark this answers is not in the document]
Comment (bowenliang123) [reply kyuubi-rest-client/src/main/java/org/apache/kyuubi/client/IRestClient.java:31]: https://github.com/apache/kyuubi/pull/4221/files#diff-0db5aaf136f196a400200a47c67d8ec6f2c42ff4ad50fe4b6d1d7784983e052aR59

File: kyuubi-rest-client/src/main/java/org/apache/kyuubi/client/BatchRestApi.java [hunk 2]
```diff
@@ -43,6 +49,16 @@ public Batch createBatch(BatchRequest request) {
...
   }
 
+  public Batch createBatch(BatchRequest request, File resourceFile) {
+    MultipartEntityBuilder entityBuilder =
+        MultipartEntityBuilder.create().setCharset(StandardCharsets.UTF_8);
+    entityBuilder.addPart(
+        "batchRequest", new StringBody(JsonUtils.toJson(request), ContentType.APPLICATION_JSON));
+    entityBuilder.addPart("resourceFile", new FileBody(resourceFile));
+    return this.getClient()
+        .post(API_BASE_PATH, entityBuilder.build(), Batch.class, client.getAuthHeader());
```
Comment (ulysses-you) [inline kyuubi-rest-client/src/main/java/org/apache/kyuubi/client/BatchRestApi.java:59]: the whole usage of entityBuilder is `resourceFile` right ? I mean there is no other code except name or file...

File: kyuubi-rest-client/src/main/java/org/apache/kyuubi/client/IRestClient.java (hunk 1 above)
[the remark this answers is not in the document]
Comment (bowenliang123) [reply kyuubi-rest-client/src/main/java/org/apache/kyuubi/client/IRestClient.java]: Prevented exposing `HttpEntity` in `BatchRestApi` by adding `Multipart` class for it . @ulysses-you @pan3793

File: kyuubi-rest-client/src/main/java/org/apache/kyuubi/client/IRestClient.java [hunk 3]
```diff
@@ -18,6 +18,7 @@
 package org.apache.kyuubi.client;
 
 import java.util.Map;
+import org.apache.kyuubi.client.api.v1.dto.MultiPart;
```
Comment (ulysses-you) [inline kyuubi-rest-client/src/main/java/org/apache/kyuubi/client/IRestClient.java:21]: nit: one more blank line

**Review by ulysses-you — APPROVED** [review]

File: kyuubi-rest-client/src/main/java/org/apache/kyuubi/client/IRestClient.java (hunk 3 above)
Comment (bowenliang123) [reply kyuubi-rest-client/src/main/java/org/apache/kyuubi/client/IRestClient.java:21]: It seems that the `dev/reformat` tool removes the blank line.

**Review by turboFei — APPROVED** [review]

**Comment by bowenliang123** [comment]
Thanks, merged to master/1.7 .

**Pull request closed without merging**